### PR TITLE
[JUJU-2030]Kill unconverted api workers 2 9

### PR DESCRIPTION
**Pull request by hmlanigan**
Split the unconverted-api-workers into two new workers and remove.

machine-setup runs setup on machines which require an API Connection. Once setup is done it can be uninstalled.

kvm-container-provisioner and lxd-container-provisioner both wait for a machine of their container type to be provisioned if the machine supports their container type. If it doesn't, the worker uninstalls. The first container provisioned of their type will cause setup for the container systems to run, then a container provisioner is started.

## QA steps

```sh
$ juju bootstrap openstack dot --constraints allocate-public-ip=true
$ juju bootstrap openstack src --constraints allocate-public-ip=true
$ juju add-model move-me 
$ juju set-model-constraints allocate-public-ip=true
$ juju add-machine lxd
$ juju add-machine kvm
$ juju add-machine 
$ juju remove-machine 0/lxd/0
$ juju add-machine lxd:0

# once settled
$ juju migrate move-me dst
$ juju switch dst:move-me

$ juju add-machine lxd:1

# checkout the juju_engine_report on each.  The kvm-container-provisioner and lxd-container-provisioner have reports on whether the container watchers are still running and if the provisioners are running. 
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1994488
https://bugs.launchpad.net/juju/+bug/1983140

### CODE REVIEW

File: cmd/jujud/agent/machine.go [hunk 1]
```diff
@@ -710,62 +707,35 @@ var (
...
-		return nil, errors.Annotate(err, "getting machine hostname")
+		return errors.Annotate(err, "getting machine hostname")
 	}
 	if err := a.recordAgentStartInformation(apiConn, hostname); err != nil {
-		return nil, errors.Annotate(err, "recording agent start information")
+		return errors.Annotate(err, "recording agent start information")
+	}
+
+	if err := a.setupContainerSupport(apiConn, logger); err != nil {
+		return err
```
Comment (tlm) [inline cmd/jujud/agent/machine.go:738]: Should we annotate this?

File: cmd/jujud/agent/machine.go [hunk 2]
```diff
@@ -785,13 +753,14 @@ func (a *MachineAgent) startAPIWorkers(apiConn api.Connection) (_ worker.Worker,
 			Commands: "[ ! -f /etc/update-manager/release-upgrades ] || sed -i '/Prompt=/ s/=.*/=never/' /etc/update-manager/release-upgrades",
 		})
 		if err != nil {
-			return nil, errors.Trace(err)
+			return errors.Trace(err)
 		}
 		if result.Code != 0 {
-			return nil, errors.New(fmt.Sprintf("cannot patch /etc/update-manager/release-upgrades: %s", result.Stderr))
+			return errors.New(fmt.Sprintf("cannot patch /etc/update-manager/release-upgrades: %s", result.Stderr))
```
Comment (tlm) [inline cmd/jujud/agent/machine.go:759]: This would be better as an fmt.Errorf call

File: cmd/jujud/agent/machine.go [hunk 3]
```diff
@@ -910,8 +879,26 @@ func (a *MachineAgent) validateMigration(apiCaller base.APICaller) error {
...
+		return errors.New("not a machine")
+	}
+	result, err := pr.Machines(mTag)
+	if err != nil {
+		return errors.Annotatef(err, "cannot load machine %s from state", mTag)
+	}
+	if len(result) != 1 {
+		return errors.Errorf("expected 1 result, got %d", len(result))
+	}
+	if errors.IsNotFound(result[0].Err) || (result[0].Err == nil && result[0].Machine.Life() == life.Dead) {
```
Comment (tlm) [inline cmd/jujud/agent/machine.go:897]: `errors.Is(err, errors.NotFound)`

File: worker/provisioner/container_initialisation.go [hunk 4]
```diff
@@ -263,11 +139,52 @@ var getContainerInitialiser = func(ct instance.ContainerType, snapChannels map[s
 	return kvm.NewContainerInitialiser()
 }
 
+func (cs *ContainerSetup) initialiseContainerProvisioner() (Provisioner, error) {
+	cs.logger.Debugf("setup provisioner for %s containers", cs.containerType)
+	if cs.managerConfig == nil {
+		return nil, errors.New("Programming error, manager config not setup")
```
Comment (tlm) [inline worker/provisioner/container_initialisation.go:145]: Should this be a notvalid ?

File: cmd/jujud/agent/engine_test.go [hunk 5]
```diff
@@ -123,15 +123,17 @@ var (
 		// "host-key-reporter", not stable, exits when done
 		"log-sender",
 		"logging-config-updater",
+		"lxd-container-provisioner",
+		"kvm-container-provisioner",
 		"machine-action-runner",
+		//"machine-setup", exits when done
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply cmd/jujud/agent/engine_test.go:129]: This one I left on purpose to explain that the machine-setup manifold isn't running when at first glance it should be.

File: cmd/jujud/agent/machine.go (hunk 1 above)
Comment (hmlanigan) [reply cmd/jujud/agent/machine.go:738]: This method can return errors from the dependency pkg. I've noticed that if wrapped, the dependency engine doesn't handle them as expected. I added logging messages to explain why we were calling ErrUninstall. Check that they make sense so we can find answers in the future.

File: cmd/jujud/agent/machine.go (hunk 3 above)
Comment (hmlanigan) [reply cmd/jujud/agent/machine.go:897]: done

File: worker/provisioner/container_initialisation.go (hunk 4 above)
Comment (hmlanigan) [reply worker/provisioner/container_initialisation.go:145]: done

File: cmd/jujud/agent/machine.go (hunk 2 above)
Comment (hmlanigan) [reply cmd/jujud/agent/machine.go:759]: very old juju 1 code, will do

File: worker/provisioner/containermanifold.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, errors.NotValidf("%q machine tag", a)
+	}
+
+	var apiCaller base.APICaller
+	if err := context.Get(cfg.APICallerName, &apiCaller); err != nil {
+		return nil, errors.Trace(err)
+	}
+	pr := apiprovisioner.NewState(apiCaller)
+
+	m, err := cfg.machineSupportsContainers(&containerShim{api: pr}, mTag)
```
Comment (tlm) [inline worker/provisioner/containermanifold.go:96]: nit can we name m to something better?

File: worker/provisioner/containermanifold.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	cfg.Logger.Debugf("%s supported containers types set as %q", mTag, types)
+
+	typeSet := set.NewStrings()
+	for _, v := range types {
+		typeSet.Add(string(v))
+	}
+	if !typeSet.Contains(string(cfg.ContainerType)) {
+		cfg.Logger.Infof("%s does not support %s containers", mTag, string(cfg.ContainerType))
```
Comment (tlm) [inline worker/provisioner/containermanifold.go:190]: s/containers/container

File: worker/provisioner/containerworker.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/core/watcher"
+)
+
+// NewContainerSetupAndProvisioner return a ContainerSetupAndProvisioner.
```
Comment (tlm) [inline worker/provisioner/containerworker.go:17]: s/return/returns

File: worker/provisioner/container_initialisation_test.go [hunk 9]
```diff
@@ -64,18 +50,22 @@ func (s *containerSetupSuite) SetUpTest(c *gc.C) {
...
 
 	s.machineLock = &fakeMachineLock{}
-	s.done = make(chan struct{})
 }
 
 var _ = gc.Suite(&containerSetupSuite{})
 
-func (s *containerSetupSuite) TestStartContainerStartsContainerProvisioner(c *gc.C) {
-	defer s.patch(c).Finish()
+func (s *containerSetupSuite) TestInitialiseContainersLXD(c *gc.C) {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply worker/provisioner/container_initialisation_test.go:57]: done

File: worker/provisioner/container_initialisation_test.go [hunk 10]
```diff
@@ -84,32 +74,42 @@ func (s *containerSetupSuite) TestStartContainerStartsContainerProvisioner(c *gc
...
+	err := cs.initialiseContainers(abort)
+	c.Assert(err, jc.ErrorIsNil)
+}
 
-	s.cleanKill(c, runner)
+func (s *containerSetupSuite) TestInitialiseContainerProvisionerLXD(c *gc.C) {
+	s.testInitialiseContainers(c, instance.KVM)
+}
+
+func (s *containerSetupSuite) TestInitialiseContainerProvisonerKVM(c *gc.C) {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply worker/provisioner/container_initialisation_test.go:88]: done

File: worker/provisioner/containermanifold.go (hunk 7 above)
Comment (hmlanigan) [reply worker/provisioner/containermanifold.go:190]: done

File: cmd/jujud/agent/agenttest/engine.go [hunk 11]
```diff
@@ -39,6 +39,14 @@ type WorkerMatcher struct {
...
 	expect    set.Strings
 	matchTime time.Time
+	optional  set.Strings
+}
+
+// AddOptionalWorkers adds workers which may or maynot start.
+// Use sparingly. Intended when these unit tests have different
+// results based on hardware.
+func (m *WorkerMatcher) AddOptionalWorkers(optional []string) {
+	m.optional = set.NewStrings(optional...)
```
Comment (tlm) [inline cmd/jujud/agent/agenttest/engine.go:49]: The keyword is here is "Add" where the operation is more of a set i.e wipe what is already there. Do we want to change the name or functionality?

**Review by tlm — APPROVED** [review]
lgtm

File: cmd/jujud/agent/agenttest/engine.go (hunk 11 above)
Comment (hmlanigan) [reply cmd/jujud/agent/agenttest/engine.go:49]: My thought was you are adding data to the matcher which cannot be set in the constructor. I didn't want to change the constructer in many places in test.

**Pull request merged**